core/src/main/scala/org/ensime/indexer/domain.scala

```diff
@@ -46,28 +49,43 @@ case class ClassName(pack: PackageName, name: String) extends FullyQualifiedName
   def fqnString =
     if (pack.path.isEmpty) name
     else ClassName.cleanupPackage(pack.fqnString + "." + name)
-
-  def internalString =
-    "L" + (if (pack.path.isEmpty) name else pack.path.mkString("/") + "/" + name) + ";"
 }
 
 object ClassName {
-  private val Root = PackageName(Nil)
-  // we consider Primitives to be ClassNames
-  private def Primitive(name: String, desc: String): ClassName = new ClassName(Root, name) {
-    override def fqnString = name
-    override def internalString = desc
+  def apply(pack: PackageName, name: String): ClassName = Impl(pack, name, internalString(pack, name))
+
+  def unapply(c: ClassName): Option[(PackageName, String)] = Some((c.pack, c.name))
+
+  private final case class Impl(pack: PackageName, name: String, internalString: String) extends ClassName {
```

Comment:
this is looking good, but I'm not sure we need `Impl`. Can you just make `ClassName` have a private constructor?, e.g. something like

`case class ClassName private(foo: String, ...)  { ... }`

and that forces everybody to go through the two-parameter `ClassName.apply`. You'll still need the two parameter `unapply` (assuming people use it).